torch/_higher_order_ops/associative_scan.py

```diff
@@ -189,25 +194,33 @@ def associative_scan_functionalize(ctx, combine_fn, input, dim):
 
 
 @associative_scan_op.py_impl(torch._C._functorch.TransformType.Vmap)
-def associative_scan_batch_rule(interpreter, input, dim, combine_fn):
-    input_ = [get_unwrapped(x) for x in input]
-    input_bdims = [maybe_get_bdim(x) for x in input]
+def associative_scan_batch_rule(interpreter, combine_fn, input, dim):
+    input_bdims = [maybe_get_bdim(x) if is_batchedtensor(x) else None for x in input]
 
     batch_size = None
     for inp, bdim in zip(input, input_bdims):
         if bdim is not None:
-            batch_size = get_unwrapped(inp).shape[bdim]
+            batch_size = (
+                get_unwrapped(inp).shape[bdim]
+                if is_batchedtensor(inp)
+                else inp.shape[bdim]
+            )
 
     assert batch_size
     input_unwrapped = []
     for x, bdim in zip(input, input_bdims):
-        unwrap = get_unwrapped(x)
+        unwrap = get_unwrapped(x) if is_batchedtensor(x) else x
         if dim is None:
             unwrap = unwrap.unsqueeze(0).expand(batch_size, *x.shape)
         else:
-            unwrap = unwrap.movedim(bdim, 0)
+            if bdim is None:
+                unwrap = unwrap.unsqueeze(0).expand(batch_size, *x.shape)
+            else:
+                unwrap = unwrap.movedim(bdim, 0)
         input_unwrapped.append(unwrap)
 
-    res = associative_scan_op(combine_fn, input_unwrapped, dim + 1)
+    with interpreter.lower():
+        res = associative_scan_op(combine_fn, input_unwrapped, dim + 1)
```

Comment:
how are vmap's in_dims, out_dims handled? Can you add some tests to convince me this is correct?

Reply:
There is now a requirement that the scan `dim` and the `in_dims` of `vmap` needs to be different. When vmapping, the batch dimension is moved to the 0-th dimension and thus the associative_scan is performed on dim+1. Then the result is concatenated together and the 0-th dimension is moved to the `out_dims` as specified for the `vmap`. I added a dedicated [testcase](https://github.com/pytorch/pytorch/pull/133013/files#diff-046f76d8f6c9786a877e46b9e686efd7a2a60a229d43cea3bc40b7461cf2b998R1236) to test this behavior. Does this convince you?